src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs

```diff
@@ -0,0 +1,89 @@
+            : base(solutionProvider, diagnosticService)
+        {
+        }
+
+        public override TextDocumentIdentifier? GetTextDocumentIdentifier(WorkspaceDocumentDiagnosticsParams request)
+            => null;
+
+        protected override WorkspaceDiagnosticReport CreateReport(TextDocumentIdentifier? identifier, VSDiagnostic[]? diagnostics, string? resultId)
+            => new WorkspaceDiagnosticReport { TextDocument = identifier, Diagnostics = diagnostics, ResultId = resultId };
+
+        protected override IProgress<WorkspaceDiagnosticReport[]>? GetProgress(WorkspaceDocumentDiagnosticsParams diagnosticsParams)
+            => diagnosticsParams.PartialResultToken;
+
+        protected override DiagnosticParams[]? GetPreviousResults(WorkspaceDocumentDiagnosticsParams diagnosticsParams)
+            => diagnosticsParams.PreviousResults;
+
+        protected override ImmutableArray<Document> GetOrderedDocuments(RequestContext context)
+        {
+            using var _ = ArrayBuilder<Document>.GetInstance(out var result);
+
+            var solution = context.Solution;
+
+            var documentTrackingService = solution.Workspace.Services.GetRequiredService<IDocumentTrackingService>();
+
+            // The active and visible docs always get priority in terms or results.
+            var activeDocument = documentTrackingService.GetActiveDocument(solution);
+            var visibleDocuments = documentTrackingService.GetVisibleDocuments(solution);
+
+            result.AddIfNotNull(activeDocument);
+            result.AddRange(visibleDocuments);
+
+            // Now, prioritize the projects related to the active/visible files.
+            AddDocumentsFromProject(activeDocument?.Project, isOpen: true);
+            foreach (var doc in visibleDocuments)
+                AddDocumentsFromProject(doc.Project, isOpen: true);
+
+            // finally, add the remainder of all documents.
+            foreach (var project in solution.Projects)
+                AddDocumentsFromProject(project, isOpen: false);
+
+            return result.Distinct().ToImmutableArray();
+
+            void AddDocumentsFromProject(Project? project, bool isOpen)
+            {
+                if (project == null)
+                    return;
+
+                // if the project doesn't necessarily have an open file in it, then only include it if the user has full
+                // solution analysis on.
+                if (!isOpen)
+                {
+                    var analysisScope = solution.Workspace.Options.GetOption(SolutionCrawlerOptions.BackgroundAnalysisScopeOption, project.Language);
+                    if (analysisScope != BackgroundAnalysisScope.FullSolution)
+                        return;
+                }
+
+                // Otherwise, if the user has an open file from this project, or FSA is on, then include all the
+                // documents from it.
+                result.AddRange(project.Documents);
```

Comment:
Could this potentially add the active / visible documents three times?  It seems like we add them to the result on line 55/56, then potentially add them again here when processing the active / visible projects, then again when processing all projects

Reply:
yes.   i will doc though.  we do a .Distinct to ensure we don't process them multiple times.  and Distinct preserves order, so adding the same doc later isn't a problem and doesn't affect earlier processing.